src/keyboard.c

```diff
@@ -132,11 +176,25 @@ static void keyboard_enter(void* data, struct wl_keyboard* wl_keyboard,
 	struct keyboard* keyboard =
 		keyboard_collection_find_wl_keyboard(collection, wl_keyboard);
 	keyboard->waiting_for_modifiers = true;
+
+	uint32_t* key;
+	wl_array_for_each(key, keys) {
+		handle_key(data, keyboard, *key + 8, XKB_KEY_DOWN);
+	}
 }
 
 static void keyboard_leave(void* data, struct wl_keyboard* wl_keyboard,
 		uint32_t serial, struct wl_surface* surface)
 {
+	struct keyboard_collection* collection = data;
+	struct keyboard* keyboard =
+		keyboard_collection_find_wl_keyboard(collection, wl_keyboard);
+
+	for (int i = 0; i < PRESSED_KEYS_MAX; i += 1) {
+		if (!keyboard->pressed_keys[i])
```

Comment:
Seems like this will miss keys if there's a gap in the array.

Seems like there are a few options here:
 * Traverse the whole array here, ignoring 0 values
 * Consolidate the array on element removal
 * Use a different algorithm

Reply:
good catch, I'v changed to continue here